### PR TITLE
Get runway stats tests

### PR DESCRIPTION
**Pull request by DelaneySteve**

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 fastapi >= 0.109.2
 requests >= 2.31.0
 uvicorn >= 0.27.1, <1.0.0
-pandas>=2.2.0
-scikit-learn>=1.4.0
+pandas >= 2.2.0
+scikit-learn >= 1.4.1.post1
```
Comment (james-zafar) [inline requirements.txt:5]: Why are we using an unstable/non-production version of sklearn?

File: data/tests/lr_model_unittest.pickle
Comment (james-zafar) [inline data/tests/lr_model_unittest.pickle:1]: I don't think we should have binaries as unit test files as it's too difficult to keep track of what's in them

File: src/api/resources/airport.py [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 
 class Airport(BaseModel):
+    iata: str
```
Comment (james-zafar) [inline src/api/resources/airport.py]: Why is this needed?

File: src/data/get_runway_stats.py
Comment (james-zafar) [inline src/data/get_runway_stats.py:1]: The logic here is very confusing to me, I think because many of the variables are poorly named. I can't tell what `runways` actually is from the name and how it's used in `__post_init__` - see questions below

File: src/data/get_runway_stats.py [hunk 3]
```diff
@@ -21,14 +21,23 @@ class RunwayStats:
...
     def __post_init__(self, runways: pd.DataFrame) -> None:
-        self._runways_stats_df = runways.map(self.count_runways)
-        self._runways_stats_df['total_runway_length'] = runways.map(self.sum_runways_len)
-
-    def count_runways(self, airport_runways: list[RunwayDict]) -> int | None:
-        return len(airport_runways) or None
-
-    def sum_runways_len(self, airport_runways: list[RunwayDict]) -> int | None:
-        return sum(runway['length_in_ft'] for runway in airport_runways) or None
+        runways_count = list(map(self.count_runways, runways['runways'], runways['iata']))
```
Comment (james-zafar) [inline src/data/get_runway_stats.py]: Why would a df of runways have an `iata` field?

Comment (james-zafar) [reply src/data/get_runway_stats.py]: Is this actually an airport rather than just the runways?

File: src/data/get_runway_stats.py [hunk 4]
```diff
@@ -21,14 +21,23 @@ class RunwayStats:
...
+        if airport_runways and any(runway['length_in_ft'] < 0 for runway in airport_runways):
+            raise ValueError('Runway length cannot be negative')
+
+    def count_runways(self, airport_runways: list[RunwayDict], iata: str) -> list[ str | int | None]:
+        self.validate_runways(airport_runways)
+        return [iata, len(airport_runways)]
+
+    def sum_runways_len(self, airport_runways: list[RunwayDict], iata: str) -> list[ str | int | None]:
+        self.validate_runways(airport_runways)
+        return [iata, sum(runway['length_in_ft'] for runway in airport_runways)]
```
Comment (james-zafar) [inline src/data/get_runway_stats.py]: Why do we need to return `iata` in every function if it's never used or modified within the function?

File: src/data/get_runway_stats.py [hunk 5]
```diff
@@ -21,14 +21,23 @@ class RunwayStats:
...
-    def count_runways(self, airport_runways: list[RunwayDict]) -> int | None:
-        return len(airport_runways) or None
-
-    def sum_runways_len(self, airport_runways: list[RunwayDict]) -> int | None:
-        return sum(runway['length_in_ft'] for runway in airport_runways) or None
+        runways_count = list(map(self.count_runways, runways['runways'], runways['iata']))
+        runways_length = list(map(self.sum_runways_len, runways['runways'],runways['iata']))
+        runways_count_df = pd.DataFrame(runways_count, columns = ['iata', 'runways'])
+        runways_length_df = pd.DataFrame(runways_length, columns = ['iata', 'total_runway_length'])
+        self._runways_stats_df = pd.merge(runways_count_df, runways_length_df, on = 'iata')
```
Comment (james-zafar) [inline src/data/get_runway_stats.py]: This seems far too complex to me. I think the same could be achieved by constructing a list of dictionaries in place that then gets converted to a dataframe at the end? Or something similar perhaps

File: tests/data_processing/test_get_runway_stats.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            raw_data = json.load(f)
+        airport_data = DataConverter(raw_data)
+        runways_df = pd.concat([airport_data.runways, airport_data.airports_df['iata']], axis=1, join='outer')
+        runway_stats = RunwayStats(runways_df)
+        runways_stats_df = runway_stats.runways_stats_df
+
+        self.assertTrue(all(isinstance(count, int) for count in runways_stats_df['runways']))
+        self.assertTrue(all(isinstance(count, int) for count in runways_stats_df['total_runway_length']))
+        self.assertIsInstance(runways_stats_df, pd.DataFrame)
+        self.assertEqual(set(runways_stats_df.columns), set(self.RUNWAYS_STATS))
```
Comment (james-zafar) [inline tests/data_processing/test_get_runway_stats.py]: As `RUNWAYS_STATS` is always converted to a set, why not store it as a set to begin with? Or better yet use `self.assertCountEqual` instead so you don't have to convert either to a set.

File: tests/data_processing/test_get_runway_stats.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                    'iata': 'YVR',
+                    'runways': []
+                }
+            ]
+        )
+        empty_runways_stats = RunwayStats(empty_runways_stats_input_df)
+        fake_output_empty = pd.DataFrame(
+            {
+            'iata': ['YVR'],
+            'runways': [0],
```
Comment (james-zafar) [inline tests/data_processing/test_get_runway_stats.py]: Why it better to label this `runway_count` or `number_of_runways` so it's clear what it is as the term `runways` seems to be used in various places to mean different things?

**Review by james-zafar — CHANGES_REQUESTED** [review]

File: data/tests/lr_model_unittest.pickle
Comment (DelaneySteve) [reply data/tests/lr_model_unittest.pickle:1]: What do you recommend using instead?

File: src/api/resources/airport.py (hunk 2 above)
Comment (DelaneySteve) [reply src/api/resources/airport.py]: Its kind of cheating but the user input gets sent through the get_runway_stats class which i've changed to require iata as an input

File: src/data/get_runway_stats.py (hunk 3 above)
Comment (DelaneySteve) [reply src/data/get_runway_stats.py]: yeah actaully that is kind of confusing its a really a df of a list of runways at different airports - i'll fix the naming

File: src/data/get_runway_stats.py (hunk 4 above)
Comment (DelaneySteve) [reply src/data/get_runway_stats.py:40]: it was the same in get_flight_stats - it gets used in the preprocessing function in the model class to merge the different runway/flight/airport stats together by airport

File: src/data/get_runway_stats.py (hunk 5 above)
Comment (DelaneySteve) [reply src/data/get_runway_stats.py]: Okay yeah I'll try that

File: tests/data_processing/test_get_runway_stats.py (hunk 7 above)
Comment (DelaneySteve) [reply tests/data_processing/test_get_runway_stats.py]: Will fix

File: src/api/resources/airport.py (hunk 2 above)
Comment (DelaneySteve) [reply src/api/resources/airport.py]: Or should I change get_runway_stats to accommodate for no iata?

Comment (james-zafar) [reply src/api/resources/airport.py]: I'd change `get_runway_stats` as we should keep only the required attributes in data structures for clarity

File: tests/data_processing/test_get_runway_stats.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        fake_output_empty = pd.DataFrame(
+            {
+            'iata': ['YVR'],
+            'runways_count': [0],
+            'total_runway_length': [0],
+            }
+        )
+        self.assertEqual(empty_runways_stats.runways_stats_df, fake_output_empty)
+
+    def test_count_runways_logic_empty_runways(self) -> None:
```
Comment (james-zafar) [inline tests/data_processing/test_get_runway_stats.py]: The name is wrong here - runways is not empty. I think this is the test above?

**Review by james-zafar — APPROVED** [review]
